in-progress/6973-refactor-base-contract-interaction.md

```diff
@@ -0,0 +1,1092 @@
+
+In a nutshell, by being more verbose in the API, we can remove a lot of complexity and make the code easier to understand and maintain; this also affords greater understanding and control over the lifecycle of contracts and transactions.
+
+Key changes:
+- the wallet is the central point of interaction to simulate/prove/send transactions instead of `BaseContractInteraction`
+- unified api for deploying contracts, deploying accounts, and calling functions
+- idempotent simulations
+- enables more sophisticated gas estimation mechanism
+
+### Major Interface Update
+
+Remove `BaseContractInteraction` and its subclasses.
+
+Define the following:
+
+```ts
+export interface UserAPI {
+  getTxExecutionRequest(userRequest: UserRequest): Promise<TxExecutionRequest>;
+  simulate(userRequest: UserRequest): Promise<SimulationOutput>;
+  read(userRequest: UserRequest): Promise<SimulationOutput>;
+  prove(userRequest: UserRequest): Promise<UserRequest>;
+  send(userRequest: UserRequest): Promise<SentTx>;
+}
+
+export type Wallet = UserAPI & AccountInterface & PXE & AccountKeyRotationInterface;
+```
+
+### Account Creation
+
+```ts
+const pxe = await createCompatibleClient(rpcUrl, debugLogger);
+const secretKey = Fr.random();
+const salt = 42;
+
+// Returns the wallet and deployment options. 
+// Does not modify PXE state.
+const { wallet: aliceWallet, deploymentArgs } = await getSchnorrWallet(pxe, secretKey, salt);
+
+const aliceContractInstance = getContractInstanceFromDeployParams(
+  SchnorrAccountContract.artifact, 
+  deploymentArgs
+);
+
+// Register the account in PXE, and wait for it to sync.
+await aliceWallet.registerAccountInPXE({ sync: true });
+
+const paymentMethod = new SomeFeePaymentMethod(
+  someCoin.address,
+  someFPC.address,
+  // the fee payment method may need a wallet,
+  // e.g. for reading the balance or producing authwits
+  aliceWallet
+);
+
+// Changes to the PXE (e.g. notes, nullifiers, auth wits, contract deployments, capsules) are not persisted.
+const { request: deployAliceAccountRequest } = await aliceWallet.simulate({
+  // easy multicall support
+  calls: [{
+    artifact: SchnorrAccountContract.artifact,
+    instance: aliceContractInstance,
+    functionName: deploymentArgs.constructorName,
+    args: deploymentArgs.constructorArgs,
+  }],
+  paymentMethod,
+  // gasSettings: undefined => automatic gas estimation. the returned `request` will have the gasSettings set.
+});
+
+// Wallet can stop here to prompt the user to confirm the gas estimation.
+
+const deployAliceAccountRequestWithProof = await aliceWallet.prove(deployAliceAccountRequest);
+
+const sentTx = await aliceWallet.send(deployAliceAccountRequestWithProof)
+
+const receipt = await sentTx.wait({
+  // permanently add the transient contracts to the wallet if the tx is successful.
+  // defaults to true.
+  registerOnSuccess: true
+});
+
+
+```
+
+### Deploy Token
```

Comment:
I'm fine with all the individual steps outlined here, but since most flows will require all these steps in this order without much change, I'd look into hiding them into something higher-level.